app/src/main/java/com/bogdan/codeforceswatcher/features/problems/redux/requests/ProblemsRequests.kt

```diff
@@ -60,8 +86,37 @@ class ProblemsRequests {
             return true
         }
 
+        private suspend fun updateDatabaseAndDispatch(newProblems: List<Problem>) {
+            val differenceProblems = mutableListOf<Problem>()
+            withContext(Dispatchers.IO) {
+                val problems = DatabaseClient.problemsDao.getAll()
+                for (problem in newProblems) {
+                    if (problems.find { it.contestId == problem.contestId && it.name == problem.name } == null) {
+                        differenceProblems.add(problem)
+                    }
+                }
+                DatabaseClient.problemsDao.insert(differenceProblems)
+            }
+            store.dispatch(Success(differenceProblems))
+        }
+
         data class Success(val problems: List<Problem>) : Action
 
         data class Failure(override val message: String?) : ToastAction
     }
+
+    class MarkProblemFavorite(val problem: Problem) : Request() {
+
+        override suspend fun execute() {
+            lateinit var newProblem: Problem
+            withContext(Dispatchers.IO) {
+                DatabaseClient.problemsDao.delete(problem)
+                newProblem = problem.copy(isFavourite = !problem.isFavourite)
+                DatabaseClient.problemsDao.insert(newProblem)
```

Comment:
I think, that insert will update a problem without deleting it. So probably you don't need to delete it.